src/main/scala/scorex/core/network/peer/PeerManager.scala

```diff
@@ -79,46 +78,66 @@ class PeerManager(settings: ScorexSettings, scorexContext: ScorexContext) extend
     NetworkUtils.isSelf(peerAddress, settings.network.bindAddress, scorexContext.externalNodeAddress)
   }
 
-  private def randomPeer(): Option[PeerInfo] = {
-    randomPeer(peerDatabase.knownPeers().values.toSeq)
+  private def isSelf(peerData: PeerData): Boolean = {
+    peerData.declaredAddress.exists(isSelf) || peerData.localAddressOpt.exists(isSelf)
   }
 
-  private def randomPeerExcluded(excludedPeers: Seq[PeerInfo]): Option[PeerInfo] = {
-    val candidates = peerDatabase.knownPeers().values.filterNot { p =>
-      excludedPeers.exists(e =>
-        e.declaredAddress == p.declaredAddress || (e.localAddress.isDefined && e.localAddress == p.localAddress)
-      )
-    }.toSeq
-
-    randomPeer(candidates)
-  }
-
-  private def randomPeer(peers: Seq[PeerInfo]): Option[PeerInfo] = {
-    if (peers.nonEmpty) Some(peers(Random.nextInt(peers.size)))
-    else None
-  }
 }
 
 object PeerManager {
 
   object ReceivableMessages {
+
     case class AddToBlacklist(remote: InetSocketAddress)
 
     // peerListOperations messages
-    case class AddOrUpdatePeer(address: InetSocketAddress,
-                               peerName: Option[String],
-                               direction: Option[ConnectionType],
-                               features: Seq[PeerFeature])
+    case class AddOrUpdatePeer(data: PeerInfo)
+
+    case class AddPeerIfEmpty(data: PeerData)
+
     case class RemovePeer(address: InetSocketAddress)
-    case object KnownPeers
-    case object RandomPeer
-    case class RandomPeers(hawMany: Int)
-    case class RandomPeerExcluding(excludedPeers: Seq[PeerInfo])
 
-    // apiInterface messages
-    case object GetAllPeers
+    /**
+      * Message to get peers from known peers map filtered by `choose` function
+      */
+    trait GetPeers[T] {
```

Comment:
```suggestion
    sealed trait GetPeers[T] {
```

Reply:
It might be useful to redefine it in concrete apps, e.g. we may want to create 'GetUTXOPeers' class in Ergo